src/features/generic_events/aggregate/event-buffer.js

```diff
@@ -0,0 +1,59 @@
+import { stringify } from '../../../common/util/stringify'
+import { MAX_PAYLOAD_SIZE } from '../constants'
+
+/**
+ * A container that keeps an event buffer and size with helper methods
+ * @typedef {Object} EventBuffer
+ * @property {number} size
+ * @property {*[]} buffer
+ */
+
+export class EventBuffer {
+  constructor () {
+    this.bytes = 0
+    this.buffer = []
+  }
+
+  /**
+   * Returns a boolean indicating whether the size and buffer contain valid data
+   * @returns {boolean}
+   */
+  get hasData () {
+    return this.buffer.length > 0 && this.bytes > 0
+  }
+
+  /**
+   * Adds an event object to the buffer while tallying size
+   * @param {Object} event the event object to add to the buffer
+   * @returns {EventBuffer} returns the event buffer for chaining
+   */
+  add (event) {
+    const size = stringify(event).length
+    if (!this.canMerge(size)) return this
+    this.buffer.push(event)
+    this.bytes += size
+    return this
+  }
+
+  /**
+   * Merges an EventBuffer into this EventBuffer
+   * @param {EventBuffer} events an EventBuffer intended to merge with this EventBuffer
+   * @param {boolean} prepend if true, the supplied events will be prepended before the events of this class
+   * @returns {EventBuffer} returns the event buffer for chaining
+   */
+  merge (eventBuffer, prepend = false) {
+    if (!this.canMerge(eventBuffer.bytes)) return
```

Comment:
Should this return `this` too?

```suggestion
    if (!this.canMerge(eventBuffer.bytes)) return this
```